docs/man/capture.1
.\" Automatically generated by Pandoc 3.1.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "TRACKER-CAPTURE" "1" "2023/10" "" "Tracker Capture Flag Manual"
.hy
.SS NAME
.PP
tracker \f[B]--capture\f[R] - Capture artifacts that were written,
executed, or found to be suspicious
.SS SYNOPSIS
.PP
tracker \f[B]--capture\f[R] <[artifact:]capture-option[=value]> \&...
.PP
tracker \f[B]--capture\f[R] <network> [\f[B]--capture\f[R]
[pcap:option1(,option2\&...)|pcap-options:option|pcap-snaplen:size]]
\&...
.SS DESCRIPTION
.PP
The \f[B]--capture\f[R] flag allows you to capture artifacts that were
written, executed, or found to be suspicious during the execution of
Tracker.
The captured artifacts will appear in the `output-path' directory.
.PP
Possible capture options:
.IP \[bu] 2
\f[B][artifact:]write[=/path/prefix*]\f[R]: Capture written files.
You can provide a filter to only capture file writes whose path starts
with a certain prefix (up to 50 characters).
Up to 3 filters can be given.
.IP \[bu] 2
\f[B][artifact:]read[=/path/prefix*]\f[R]: Capture read files.
You can provide a filter to only capture file reads whose path starts
with a certain prefix (up to 50 characters).
Up to 3 filters can be given.
.IP \[bu] 2
\f[B][artifact:]exec\f[R]: Capture executed files.
.IP \[bu] 2
\f[B][artifact:]module\f[R]: Capture loaded kernel modules.
.IP \[bu] 2
\f[B][artifact:]bpf\f[R]: Capture loaded BPF programs bytecode.
.IP \[bu] 2
\f[B][artifact:]mem\f[R]: Capture memory regions that had write+execute
(w+x) protection and then changed to execute (x) only.
.IP \[bu] 2
\f[B][artifact:]network\f[R]: Capture network traffic.
Only TCP/UDP/ICMP protocols are currently supported.
.SS File Capture Filters
.PP
Files captured upon read/write can be filtered to catch only specific IO
operations.
The different filter types have a logical `AND' between them but a
logical `OR' between filters of the same type.
The filter format is as follows:
<read/write>:<filter-type>=<filter-value>
.PP
Filter types:
.IP \[bu] 2
\f[B]path\f[R]: A filter for the file path prefix (up to 50 characters).
Up to 3 filters can be given.
Identical to using `<read/write>=/path/prefix*'.
.IP \[bu] 2
\f[B]type\f[R]: A file type from the following options: `regular',
`pipe', and `socket'.
.IP \[bu] 2
\f[B]fd\f[R]: The file descriptor of the file.
Can be one of the three standards: `stdin', `stdout', and `stderr'.
.SS Network Capture Notes
.IP \[bu] 2
Pcap Files:
.RS 2
.IP \[bu] 2
If you only specify \f[B]--capture network\f[R], you will have a single
file with all network traffic.
.IP \[bu] 2
You can use \f[B]pcap:xxx,yyy\f[R] to have more than one pcap file,
split by different means.
.RE
.IP \[bu] 2
Pcap Options:
.RS 2
.IP \[bu] 2
If you do not specify \f[B]pcap-options\f[R] (or set to none), you will
capture ALL network traffic into your pcap files.
.IP \[bu] 2
If you specify \f[B]pcap-options:filtered\f[R], events being traced will
define what network traffic will be captured.
.RE
.IP \[bu] 2
Snap Length:
.RS 2
.IP \[bu] 2
If you do not specify a snaplen, the default is headers only (incomplete
packets in tcpdump).
.IP \[bu] 2
If you specify \f[B]max\f[R] as snaplen, you will get the full contents
of each packet (pcap files will be large).
.IP \[bu] 2
If you specify \f[B]headers\f[R] as snaplen, you will only get L2/L3
headers in captured packets.
.IP \[bu] 2
If you specify \f[B]headers\f[R] but trace for \f[B]net_packet_dns\f[R]
events, the L4 DNS header will be captured.
.IP \[bu] 2
If you specify \f[B]headers\f[R] but trace for \f[B]net_packet_http\f[R]
events, only L2/L3 headers will be captured.
.RE
.SS EXAMPLES
.SS File capture
.IP \[bu] 2
To capture executed files into the default output directory, use the
following flag:
.RS 2
.IP
.nf
\f[C]
--capture exec
\f[R]
.fi
.RE
.IP \[bu] 2
To capture executed files into a specific directory, clear the directory
before starting, use the following flags:
.RS 2
.IP
.nf
\f[C]
--capture exec --capture dir:/my/dir --capture clear-dir
\f[R]
.fi
.RE
.IP \[bu] 2
To capture files that were written into anywhere under /usr/bin/ or
/etc/, use the following flags:
.RS 2
.IP
.nf
\f[C]
--capture write=/usr/bin/* --capture write=/etc/*
\f[R]
.fi
.RE
.IP \[bu] 2
To capture file writes to socket files that are the `stdout' of the
writing process, use the following flag:
.RS 2
.IP
.nf
\f[C]
--capture write:type=socket --capture write:fd=stdout
\f[R]
.fi
.RE
.SS Network Capture
.IP \[bu] 2
To capture network traffic, use the following flag:
.RS 2
.IP
.nf
\f[C]
--capture network
\f[R]
.fi
.RE
.IP \[bu] 2
To capture network traffic and save separate pcap files for processes
and commands, use the following flag:
.RS 2
.IP
.nf
\f[C]
--capture network --capture pcap:process,command
\f[R]
.fi
.RE
.IP \[bu] 2
To capture network traffic and save pcap files containing only
traced/filtered packets, use the following flag:
.RS 2
.IP
.nf
\f[C]
--capture network --capture pcap-options:filtered
\f[R]
.fi
.RE
.IP \[bu] 2
To capture network traffic and set the captured payload from each packet
to 1KB, use the following flag:
.RS 2
.IP
.nf
\f[C]
--capture network --capture pcap-snaplen:1kb
\f[R]
.fi
.RE
.IP \[bu] 2
To capture network traffic and save pcap files for containers and
commands, use the following flag:
.RS 2
.IP
.nf
\f[C]
--capture network --capture pcap:container,command
\f[R]
.fi
.RE


pkg/ebpf/capture.go
package ebpf

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"os"
	"path"
	"strings"

	"github.com/khulnasoft-lab/tracker/pkg/bufferdecoder"
	"github.com/khulnasoft-lab/tracker/pkg/errfmt"
	"github.com/khulnasoft-lab/tracker/pkg/logger"
	"github.com/khulnasoft-lab/tracker/pkg/utils"
)

func (t *Tracker) processFileCaptures(ctx context.Context) {
	logger.Debugw("Starting processFileCaptures go routine")
	defer logger.Debugw("Stopped processFileCaptures go routine")

	const (
		// stat_S_IFMT uint32 = 0170000 // bit mask for the file type bit field
		_S_IFSOCK uint32 = 0140000 // socket
		_S_IFLNK  uint32 = 0120000 // symbolic link
		_S_IFREG  uint32 = 0100000 // regular file
		_S_IFBLK  uint32 = 0060000 // block device
		_S_IFDIR  uint32 = 0040000 // directory
		_S_IFCHR  uint32 = 0020000 // character device
		_S_IFIFO  uint32 = 0010000 // FIFO
	)

	for {
		select {
		case dataRaw := <-t.fileCapturesChannel:
			if len(dataRaw) == 0 {
				continue
			}
			ebpfMsgDecoder := bufferdecoder.New(dataRaw)
			var meta bufferdecoder.ChunkMeta
			appendFile := false

			err := ebpfMsgDecoder.DecodeChunkMeta(&meta)
			if err != nil {
				t.handleError(err)
				continue
			}

			if meta.Size <= 0 {
				t.handleError(errfmt.Errorf("invalid chunk size: %d", meta.Size))
				continue
			}

			if ebpfMsgDecoder.BuffLen() < int(meta.Size) {
				t.handleError(errfmt.Errorf("chunk too large: %d", meta.Size))
				continue
			}

			containerId := t.containers.GetCgroupInfo(meta.CgroupID).Container.ContainerId
			if containerId == "" {
				containerId = "host"
			}
			pathname := containerId
			if err := utils.MkdirAtExist(t.OutDir, pathname, 0755); err != nil {
				t.handleError(err)
				continue
			}
			filename := ""
			metaBuffDecoder := bufferdecoder.New(meta.Metadata[:])
			var kernelModuleMeta bufferdecoder.KernelModuleMeta
			var bpfObjectMeta bufferdecoder.BpfObjectMeta
			if meta.BinType == bufferdecoder.SendVfsWrite || meta.BinType == bufferdecoder.SendVfsRead {
				var vfsMeta bufferdecoder.VfsFileMeta
				err = metaBuffDecoder.DecodeVfsFileMeta(&vfsMeta)
				if err != nil {
					t.handleError(err)
					continue
				}
				if vfsMeta.Mode&_S_IFSOCK == _S_IFSOCK || vfsMeta.Mode&_S_IFCHR == _S_IFCHR || vfsMeta.Mode&_S_IFIFO == _S_IFIFO {
					appendFile = true
				}
				var operation string
				if meta.BinType == bufferdecoder.SendVfsRead {
					operation = "read"
				} else {
					operation = "write"
				}
				if vfsMeta.Pid == 0 {
					filename = fmt.Sprintf(
						"%s.dev-%d.inode-%d",
						operation,
						vfsMeta.DevID,
						vfsMeta.Inode,
					)
				} else { // Only applies for write to /dev/null
					filename = fmt.Sprintf(
						"%s.dev-%d.inode-%d.pid-%d",
						operation,
						vfsMeta.DevID,
						vfsMeta.Inode,
						vfsMeta.Pid,
					)
				}
			} else if meta.BinType == bufferdecoder.SendMprotect {
				var mprotectMeta bufferdecoder.MprotectWriteMeta
				err = metaBuffDecoder.DecodeMprotectWriteMeta(&mprotectMeta)
				if err != nil {
					t.handleError(err)
					continue
				}
				// note: size of buffer will determine maximum extracted file size! (as writes from kernel are immediate)
				if t.config.Output.RelativeTime {
					// To get the monotonic time since tracker was started, we have to subtract the start time from the timestamp.
					mprotectMeta.Ts -= t.startTime
				} else {
					// To get the current ("wall") time, we add the boot time into it.
					mprotectMeta.Ts += t.bootTime
				}
				filename = fmt.Sprintf("bin.pid-%d.ts-%d", mprotectMeta.Pid, mprotectMeta.Ts)
			} else if meta.BinType == bufferdecoder.SendKernelModule {
				err = metaBuffDecoder.DecodeKernelModuleMeta(&kernelModuleMeta)
				if err != nil {
					t.handleError(err)
					continue
				}
				filename = "module"
				if kernelModuleMeta.DevID != 0 {
					filename = fmt.Sprintf("%s.dev-%d", filename, kernelModuleMeta.DevID)
				}
				if kernelModuleMeta.Inode != 0 {
					filename = fmt.Sprintf("%s.inode-%d", filename, kernelModuleMeta.Inode)
				}
				if kernelModuleMeta.Pid != 0 {
					filename = fmt.Sprintf("%s.pid-%d", filename, kernelModuleMeta.Pid)
				}
			} else if meta.BinType == bufferdecoder.SendBpfObject {
				err = metaBuffDecoder.DecodeBpfObjectMeta(&bpfObjectMeta)
				if err != nil {
					t.handleError(err)
					continue
				}
				bpfName := string(bytes.TrimRight(bpfObjectMeta.Name[:], "\x00"))
				filename = fmt.Sprintf("bpf.name-%s", bpfName)
				if bpfObjectMeta.Pid != 0 {
					filename = fmt.Sprintf("%s.pid-%d", filename, bpfObjectMeta.Pid)
				}
				filename = fmt.Sprintf("%s.%d", filename, bpfObjectMeta.Rand)
			} else {
				t.handleError(errfmt.Errorf("unknown binary type: %d", meta.BinType))
				continue
			}

			fullname := path.Join(pathname, filename)

			f, err := utils.OpenAt(t.OutDir, fullname, os.O_CREATE|os.O_WRONLY, 0640)
			if err != nil {
				t.handleError(err)
				continue
			}
			if appendFile {
				if _, err := f.Seek(0, io.SeekEnd); err != nil {
					if err := f.Close(); err != nil {
						t.handleError(err)
					}
					t.handleError(err)
					continue
				}
			} else {
				if _, err := f.Seek(int64(meta.Off), io.SeekStart); err != nil {
					if err := f.Close(); err != nil {
						t.handleError(err)
					}
					t.handleError(err)
					continue
				}
			}

			dataBytes, err := bufferdecoder.ReadByteSliceFromBuff(ebpfMsgDecoder, int(meta.Size))
			if err != nil {
				if err := f.Close(); err != nil {
					t.handleError(err)
				}
				t.handleError(err)
				continue
			}
			if _, err := f.Write(dataBytes); err != nil {
				if err := f.Close(); err != nil {
					t.handleError(err)
				}
				t.handleError(err)
				continue
			}
			if err := f.Close(); err != nil {
				t.handleError(err)
				continue
			}
			// Rename the file to add hash when last chunk was received
			if meta.BinType == bufferdecoder.SendKernelModule && uint32(meta.Size)+uint32(meta.Off) == kernelModuleMeta.Size {
				fileHash, _ := t.computeOutFileHash(fullname)
				err := utils.RenameAt(t.OutDir, fullname, t.OutDir, fullname+"."+fileHash)
				if err != nil {
					t.handleError(err)
					continue
				}
			} else if meta.BinType == bufferdecoder.SendBpfObject && (uint32(meta.Size)+uint32(meta.Off)) == bpfObjectMeta.Size {
				fileHash, _ := t.computeOutFileHash(fullname)
				// Delete the random int used to differentiate files
				dotIndex := strings.LastIndex(fullname, ".")
				err := utils.RenameAt(t.OutDir, fullname, t.OutDir, fullname[:dotIndex]+"."+fileHash)
				if err != nil {
					t.handleError(err)
					continue
				}
			}

		case lost := <-t.lostCapturesChannel:
			if err := t.stats.LostWrCount.Increment(lost); err != nil {
				logger.Errorw("Incrementing lost capture count", "error", err)
			}
			logger.Warnw(fmt.Sprintf("Lost %d capture events", lost))

		case <-ctx.Done():
			return
		}
	}
}
